### PR TITLE
 validate capabilities in alias config

### PR DESCRIPTION
**Pull request by Sonali-More-Xandr**
This PR adds validation checks for alias capabilities. This check makes sure that values defined for alias capabilities are subset of parent bidder capabilities values.

### CODE REVIEW

File: config/bidderinfo.go [hunk 1]
```diff
@@ -338,14 +338,16 @@ func validateAdapterEndpoint(endpoint string, bidderName string, errs []error) [
 	return errs
 }
 
-func validateInfo(info BidderInfo, bidderName string) error {
+func validateInfo(info BidderInfo, infos BidderInfos, bidderName string) error {
```
Comment (SyntaxNode) [inline config/bidderinfo.go:341]: Please differentiate the names `info` and `infos` more. The single character difference is too confusing.

File: config/bidderinfo.go [hunk 2]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
 	return nil
 }
 
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
+	if len(bidder.AliasOf) > 0 {
+		if parentBidder, ok := infos[bidder.AliasOf]; ok {
+			if parentBidder.Capabilities == nil && bidder.Capabilities != nil {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
```
Comment (SyntaxNode) [inline config/bidderinfo.go:365]: You should use errors.New here. There is no advantage to using fmt.Errorf if there is no formatting in the error string.

File: config/bidderinfo.go [hunk 3]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+			if (parentBidder.Capabilities.App == nil && bidder.Capabilities.App != nil) || (parentBidder.Capabilities.App != nil && bidder.Capabilities.App == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			} else if bidder.Capabilities.App != nil {
+				return validateAliasPlatformInfo(parentBidder.Capabilities.App, bidder.Capabilities.App)
+			}
+		} else {
+			return fmt.Errorf("bidder: %s not found for an alias: %s", bidder.AliasOf, bidderName)
+		}
+	}
+	return nil
```
Comment (SyntaxNode) [inline config/bidderinfo.go:381]: Deeply nested `if` statements are difficult to follow. Please reverse the logic of your checks, something like:

```
if len(bidder.AliasOf) == 0 {
  return nil
}

parentBidder, parentFound := infos[bidder.AliasOf]
if !parentFound {
  return fmt.Errorf("bidder: %s not found for an alias: %s", bidder.AliasOf, bidderName)
}

....
```

File: config/bidderinfo.go [hunk 4]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
 
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
+	if len(bidder.AliasOf) > 0 {
+		if parentBidder, ok := infos[bidder.AliasOf]; ok {
+			if parentBidder.Capabilities == nil && bidder.Capabilities != nil {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			}
+			if (parentBidder.Capabilities.Site == nil && bidder.Capabilities.Site != nil) || (parentBidder.Capabilities.Site != nil && bidder.Capabilities.Site == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			} else if bidder.Capabilities.Site != nil {
```
Comment (SyntaxNode) [inline config/bidderinfo.go:369]: Nitpick: There is no benefit to using an `else if` when the `if` condition results in a function exit.

File: config/bidderinfo.go [hunk 5]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+		} else {
+			return fmt.Errorf("bidder: %s not found for an alias: %s", bidder.AliasOf, bidderName)
+		}
+	}
+	return nil
+}
+
+func validateAliasPlatformInfo(parentInfo *PlatformInfo, aliasInfo *PlatformInfo) error {
+	parentInfoMap := make(map[openrtb_ext.BidType]struct{})
+	for _, info := range parentInfo.MediaTypes {
```
Comment (SyntaxNode) [inline config/bidderinfo.go:386]: If you pass `parentInfo` and `aliasInfo` by reference, you need to handle the nil case.

File: config/bidderinfo.go [hunk 6]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+		parentInfoMap[info] = struct{}{}
+	}
+
+	for _, info := range aliasInfo.MediaTypes {
+		if _, found := parentInfoMap[info]; !found {
+			return fmt.Errorf("alias MediaTypes should be a subset of parent bidder MediaTypes")
```
Comment (SyntaxNode) [inline config/bidderinfo.go:392]: No need to use fmt.Errorf here.

File: config/bidderinfo.go (hunk 2 above)
Comment (AlexBVolcy) [reply config/bidderinfo.go:365]: Can you also update tests to include coverage for this line?

File: config/bidderinfo.go [hunk 7]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+			}
+			if (parentBidder.Capabilities.Site == nil && bidder.Capabilities.Site != nil) || (parentBidder.Capabilities.Site != nil && bidder.Capabilities.Site == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			} else if bidder.Capabilities.Site != nil {
+				return validateAliasPlatformInfo(parentBidder.Capabilities.Site, bidder.Capabilities.Site)
+			}
+			if (parentBidder.Capabilities.App == nil && bidder.Capabilities.App != nil) || (parentBidder.Capabilities.App != nil && bidder.Capabilities.App == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			} else if bidder.Capabilities.App != nil {
+				return validateAliasPlatformInfo(parentBidder.Capabilities.App, bidder.Capabilities.App)
```
Comment (AlexBVolcy) [inline config/bidderinfo.go:375]: Can you also update tests to include coverage for these lines?

File: config/bidderinfo.go (hunk 1 above)
Comment (AlexBVolcy) [reply config/bidderinfo.go:341]: Yeah I agree with this

File: config/bidderinfo.go [hunk 8]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+		}
+	}
+	return nil
+}
+
+func validateAliasPlatformInfo(parentInfo *PlatformInfo, aliasInfo *PlatformInfo) error {
+	parentInfoMap := make(map[openrtb_ext.BidType]struct{})
+	for _, info := range parentInfo.MediaTypes {
+		parentInfoMap[info] = struct{}{}
+	}
```
Comment (onkarvhanumante) [inline config/bidderinfo.go:388]: `parentInfoMap` holds `mediaTypes`. how about renaming `parentInfoMap` to `parentMediaTypes`

File: config/bidderinfo.go [hunk 9]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
+	if len(bidder.AliasOf) > 0 {
+		if parentBidder, ok := infos[bidder.AliasOf]; ok {
+			if parentBidder.Capabilities == nil && bidder.Capabilities != nil {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			}
+			if (parentBidder.Capabilities.Site == nil && bidder.Capabilities.Site != nil) || (parentBidder.Capabilities.Site != nil && bidder.Capabilities.Site == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			} else if bidder.Capabilities.Site != nil {
+				return validateAliasPlatformInfo(parentBidder.Capabilities.Site, bidder.Capabilities.Site)
```
Comment (onkarvhanumante) [inline config/bidderinfo.go:370]: can pass  `*parentBidder.Capabilities.Site` `*bidder.Capabilities.Site` to `validateAliasPlatformInfo`

File: config/bidderinfo.go [hunk 10]
```diff
@@ -356,6 +358,44 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
 }
 
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
+	if len(bidder.AliasOf) > 0 {
+		if parentBidder, ok := infos[bidder.AliasOf]; ok {
+			if parentBidder.Capabilities == nil && bidder.Capabilities != nil {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
+			}
+			if (parentBidder.Capabilities.Site == nil && bidder.Capabilities.Site != nil) || (parentBidder.Capabilities.Site != nil && bidder.Capabilities.Site == nil) {
+				return fmt.Errorf("alias capabilities should be a subset of parent bidder capabilities")
```
Comment (onkarvhanumante) [inline config/bidderinfo.go:368]: how about  including site capabilities in error message? makes more clearer what capabilities is not subset

```
return fmt.Errorf("site capabilities of alias should be match or subset of parent bidder")
```

File: config/bidderinfo.go (hunk 2 above)
Comment (onkarvhanumante) [inline config/bidderinfo.go:365]: how about changing error message to
```
capabilities of alias should be match or subset of parent bidder
```

File: config/bidderinfo_test.go [hunk 11]
```diff
@@ -518,6 +518,7 @@ func TestBidderInfoValidationNegative(t *testing.T) {
 			[]error{
 				errors.New("bidder: bidderB cannot be an alias of an alias: bidderA"),
 				errors.New("bidder: bidderC not found for an alias: bidderB"),
+				errors.New("bidder: bidderC not found for an alias: bidderB"),
```
Comment (SyntaxNode) [inline config/bidderinfo_test.go:521]: Is there now two errors for the alias not found error?

**Review by SyntaxNode — COMMENTED** [review]
Thank you for following Go conventions on error messages (don't start with an upper case letter, don't end with punctuation)

File: config/bidderinfo.go (hunk 2 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: * Updated error messages to list bidders with error as part of [96e960b](https://github.com/prebid/prebid-server/pull/2996/commits/96e960b0bdee0e8670ee19308e4f2c42104939be) so keeping `fmt.Errorf`
* Added a test case as part of [1b545f5](https://github.com/prebid/prebid-server/pull/2996/commits/1b545f54db7d90138f35ccf70b7e9744301fd917)

File: config/bidderinfo.go (hunk 9 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: Updated in [ba622f5](https://github.com/prebid/prebid-server/pull/2996/commits/ba622f5d216bfead0ca7beea29b32ba27eb552d6)

File: config/bidderinfo.go (hunk 8 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: Renamed in [4ebb3d1](https://github.com/prebid/prebid-server/pull/2996/commits/4ebb3d181c9962fa14a44ec869f41dd2135eee65)

File: config/bidderinfo_test.go (hunk 11 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo_test.go:521]: yes, one from `validateAliases` and another from `validateAliasCapabilities`

File: config/bidderinfo.go (hunk 2 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: Error message are updated as part of [96e960b](https://github.com/prebid/prebid-server/pull/2996/commits/96e960b0bdee0e8670ee19308e4f2c42104939be)

File: config/bidderinfo.go (hunk 1 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: updated in [1081cc7](https://github.com/prebid/prebid-server/pull/2996/commits/1081cc7870716a3bfb638a8d225726b8115a9a5e)

**Review by AlexBVolcy — APPROVED** [review]

**Review by onkarvhanumante — DISMISSED** [review]

**Review by AlexBVolcy — APPROVED** [review]

File: config/bidderinfo.go [hunk 12]
```diff
@@ -356,6 +358,50 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
 
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
+	if len(bidder.AliasOf) == 0 {
+		return nil
+	}
+
+	parentBidder, parentFound := infos[bidder.AliasOf]
+	if !parentFound {
+		return fmt.Errorf("bidder: %s not found for an alias: %s", bidder.AliasOf, bidderName)
+	}
```
Comment (gargcreation1992) [inline config/bidderinfo.go:369]: I think we should just validate the capabilities part here. This check is already covered in https://github.com/prebid/prebid-server/blob/92211a53da0a3a1dcf80c64c792f36265d88ffcb/config/bidderinfo.go#L331-L342 
At this point, that check should have already been called. So, no need for re-check and safely use `parentBidder := infos[bidder.AliasOf]`

File: config/bidderinfo.go [hunk 13]
```diff
@@ -356,6 +358,50 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
 	return nil
 }
 
+func validateAliasCapabilities(bidder BidderInfo, infos BidderInfos, bidderName string) error {
```
Comment (gargcreation1992) [inline config/bidderinfo.go:361]: We are validating aliasBidder capabilities here. I think it is more intuitive to use `aliasBidder` or `aliasBidderInfo` in function arguments. 
```
func validateAliasCapabilities(aliasBidderInfo BidderInfo, infos BidderInfos, bidderName string) error {
```

File: config/bidderinfo.go [hunk 14]
```diff
@@ -356,6 +358,50 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
+			return validateAliasPlatformInfo(*parentBidder.Capabilities.App, *bidder.Capabilities.App, bidderName, bidder.AliasOf)
+		}
+
+		return fmt.Errorf("capabilities for alias: %s should be a subset of capabilities for parent bidder: %s", bidderName, bidder.AliasOf)
+	}
+
+	return nil
+}
+
+func validateAliasPlatformInfo(parentInfo PlatformInfo, aliasInfo PlatformInfo, bidderName string, parentBidderName string) error {
```
Comment (gargcreation1992) [inline config/bidderinfo.go:390]: nitpick - might wanna consider renaming this function to `isAliasPlatformInfoSubsetOfParent`

**Review by gargcreation1992 — COMMENTED** [review]
few nitpicks you might wanna consider.

**Review by onkarvhanumante — APPROVED** [review]

File: config/bidderinfo.go [hunk 15]
```diff
@@ -403,36 +405,32 @@ func validateMaintainer(info *MaintainerInfo, bidderName string) error {
...
-	if len(bidder.AliasOf) == 0 {
-		return nil
-	}
-
-	parentBidder, parentFound := infos[bidder.AliasOf]
+func validateAliasCapabilities(aliasBidderInfo BidderInfo, infos BidderInfos, bidderName string) error {
+	parentBidder, parentFound := infos[aliasBidderInfo.AliasOf]
 	if !parentFound {
-		return fmt.Errorf("bidder: %s not found for an alias: %s", bidder.AliasOf, bidderName)
+		return fmt.Errorf("bidder: %s not found for an alias: %s", aliasBidderInfo.AliasOf, bidderName)
```
Comment (AlexBVolcy) [inline config/bidderinfo.go:411]: Can you add a test to cover this error case?

**Review by AlexBVolcy — COMMENTED** [review]
Looks good, left one test coverage comment

**Review by onkarvhanumante — APPROVED** [review]

File: config/bidderinfo_test.go [hunk 16]
```diff
@@ -928,6 +928,303 @@ func TestBidderInfoValidationNegative(t *testing.T) {
 				errors.New("The endpoint: incorrect for bidderB is not a valid URL"),
 			},
 		},
+		{
```
Comment (gargcreation1992) [inline config/bidderinfo_test.go:931]: Could we please write one test case that passes?

**Review by gargcreation1992 — COMMENTED** [review]
Requesting to please write one +ve test case.

File: config/bidderinfo_test.go (hunk 16 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo_test.go:931]: Positive test is already covered in [TestBidderInfoValidationPositive](https://github.com/prebid/prebid-server/blob/master/config/bidderinfo_test.go#L499)

File: config/bidderinfo.go (hunk 15 above)
Comment (Sonali-More-Xandr) [reply config/bidderinfo.go]: Added in [0db8d10](https://github.com/prebid/prebid-server/pull/2996/commits/0db8d100c06228c2360c7b19b4500a3dbb8a4c54)

**Review by gargcreation1992 — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Pull request merged**